lua/lsp/init.lua

```diff
@@ -69,12 +69,18 @@ end
 
 function M.setup(lang)
   local lsp = lvim.lang[lang].lsp
-  if require("utils").check_lsp_client_active(lsp.provider) then
-    return
+
+  -- normalize the lsp config to be a table of providers
+  if lsp["provider"] ~= nil then
```

Comment:
If you want to add support to multiple providers, please change the config structure to show that it supports this.

Reply:
My idea was to make this feature optional and not do any changes to `lua/default-config.lua`, as it might break the setup for some users on `rolling`.

I can add my php example to `utils/installer/config.example.lua` if that's what you mean.

Reply:
No I mean that we shouldn't have some kind of adapter, we should adapt the structure of the configuration to our needs, and it should reflect on Lunarvim's capabilities.

Reply:
So probably it should be `providers = {...}`

Reply:
> No I mean that we shouldn't have some kind of adapter, we should adapt the structure of the configuration to our needs, and it should reflect on Lunarvim's capabilities.

Ok, I agree

> So probably it should be `providers = {...}`

What exactly do you think should be inside the `{...}`? Just the names of providers, or setup functions too?

Reply:
I propose a structure like this:
```
lvim.lang.php.lsp = {
  intelephense = {
    setup = { },
    },
  },
  phpactor = {
    setup = {
    },
  },
}
```

Basically the one I had in my example, but we remove the redundant provider key.

I can refactor the default config to be structured like this, if you're sure that's ok.

Reply:
Hello, I am still interested in contributing this as well as refactor the layout of default configs for lsp, I just want to agree on some specific structure, so I don't have to redo it.

Reply:
The structure should be

``` lua
lvim.lang.php.lsp = {
  providers = {
    {
      name="provider_name",
      setup = {}
    }
  }
}
```

But I don't think configuring multiple LSPs just work out of the box, for instance how does it work if both provide formatting capabilities ?

Reply:
Ok, I agree with your suggested structure.

> But I don't think configuring multiple LSPs just work out of the box, for instance how does it work if both provide formatting capabilities ?

You can see in my example config in the first post I am passing custom handlers to make sure there is only one hover box when I press `Shift+k`. In my experience some of the stuff you disable via `capabilities` and some via `handlers`, but I think it should be up to user to decide which language server they want to use for each functionality. It is up to us, however, to make sure that there is a way to disable the functionalities and it's well documented. I will update sample config with an example.

Reply:
@Tastyep Hello, I have updated the default config with new `lsp.providers` structure, removed normalization code and made it somewhat more streamlined overall.

Also updated the example configs with infos on how to add 2 providers and added a note about disabling capabilities via handlers. The `["textDocument/hover"] = function() end,` trick works for pretty much anything I have tested:
* textDocument/formatting
* textDocument/hover
* textDocument/publishDiagnostics
* textDocument/definition
* textDocument/declaration

One thing I didn't find a way to remove duplication for is `textDocument/completion`. The handler trick does nothing for it, and neither does setting capability to `nil` or false. Same for overwriting a callback. I am under impression that this should be resolved via `compe`, but I'm not sure there is a way to disable LSP source for just one particular provider.

Really hope this gets merged before any other changes to default configs are made, resolving configs there won't be pleasant.